<a href="https://colab.research.google.com/github/lustraka/puzzles/blob/main/AoC2021/AoC_13.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Advent of Code Puzzles
[Advent of Code 2021](https://adventofcode.com/2021) | [reddit/adventofcode](https://www.reddit.com/r/adventofcode/)

In [7]:
import requests
import pandas as pd
import numpy as np
import re
from collections import Counter
path = 'https://raw.githubusercontent.com/lustraka/puzzles/main/AoC2021/data/'

## [Day 15](https://adventofcode.com/2021/day/15): Chiton
### Part I
- **Unknown**: The lowest risk of a path from the top left to the bottom right.
- **Data**: A map of risk level.
- **Condition**:
  - Add up the risk levels of each position you enter.
  - You cannot move diagonally.
  - We don't need to store the shortest path.

Dependecies to consider:
```python
from collections import Counter, defaultdict, deque
from heapq import heappop, heappush
```

In [8]:
example = """1163751742
1381373672
2136511328
3694931569
7463417111
1319128137
1359912421
3125421639
1293138521
2311944581"""

In [64]:
from collections import defaultdict
def parse(input):
  # Initialize an array
  map = [list(line) for line in input.split('\n')]
  map = np.array(map).astype(int)
  # Transform the array into an default dict
  mapdict = defaultdict(lambda: np.inf)
  for i in range(map.shape[0]):
    for j in range(map.shape[1]):
      mapdict[(i,j)] = map[i,j]
  # Return the default dict, the start position and the end position
  return mapdict, (0, 0), tuple(np.array(map.shape)-1)

map, start, end = parse(example)
len(map), start, end

(100, (0, 0), (9, 9))

- **Plan** (Implement [Dijkstra's algorithm](https://www.wikiwand.com/en/Dijkstra%27s_algorithm)):
  - Create the vertex set `Q`
  - Initialize the vertex `u` with the `start`
  - Initialize the default dict `dist` that contains the current distances from the start to other vertices.
  - For each neighbour `v` of `u`:
    - calculate the distance `alt` to `v` through `u`
    - if `alt` < `dist[v]` update `dist[v]`
  - Repeat for the entire map


In [96]:
def solve_part1(input):
  map, start, end = parse(input)
  # Create vertex set Q
  Q = list(map.keys())
  # Initialize the vertex u
  u = start
  # Initialize map of distances from start dist
  dist = defaultdict(lambda: np.inf)
  for k, _ in map.items():
    dist[k] = np.inf
  dist[start] = 0
  for i in range(end[0]+1):
    for j in range(end[1]+1):
      # Identify neighbours not yet visited
      neighbours = [(i+1,j),(i-1,j),(i,j+1),(i,j-1)]
      # Compute and update distances
      for v in neighbours:
        alt = dist[i,j] + map[v]
        if alt < dist[v]:
          dist[v] = alt
      Q.remove((i,j))
  return dist, dist[end]

dist, risk = solve_part1(example)
risk

40

In [97]:
r = requests.get(path+'AoC2021_15.txt')
dist, risk = solve_part1(r.text[:-1])
risk

402

`402` : That's not the right answer; your answer is too high.

In [104]:
frag = []
for i in range(97,100):
  for j in range(97,100):
    frag.append(dist[i,j])
np.array(frag).reshape(3,3)

array([[390, 395, 397],
       [391, 392, 397],
       [396, 393, 402]])

In [105]:
frag = []
for i in range(5):
  for j in range(5):
    frag.append(dist[i,j])
np.array(frag).reshape(5,5)

array([[ 0,  5,  7,  9, 13],
       [ 7,  6,  7, 16, 14],
       [10,  7,  9, 16, 16],
       [ 9,  8, 16, 16, 17],
       [15, 13, 14, 15, 20]])

This works but the value of (9,9) is 41 (not 40). Algorithm get lost in a maze somehow...
```python
def dijkstra_iter(map, dist, u, end):
  # Identify neighbours not yet visited
  neighbours = [v for v in [(u[0]+1,u[1]),(u[0]-1,u[1]),(u[0],u[1]+1),(u[0],u[1]-1)] if v in Q]
  # Compute and update distances
  for v in neighbours:
    alt = dist[u] + map[v]
    if alt < dist[v]:
      dist[v] = alt
  # Search the next node
  if neighbours:
    min = neighbours[0]
  else:
    min = Q[0]
  for v in neighbours[1:]:
    if dist[v] < dist[min]:
      min = v
  # Remove u from Q and continue with min
  Q.remove(u)
  if min in Q:
    u = min
  else:
    u = Q[0]
  print(u, dist[u])
  return u

while Q:
  u = dijkstra_iter(map,dist,u,end)
  if u == end:
    print(dist[u])
    break
```

### Part II
- **Unknown**: 
- **Data**: 
- **Condition**:
  - 
  - 
- **Plan**:
  - 


## [Day 14](https://adventofcode.com/2021/day/14): Extended Polymerization
### Part I
- **Unknown**: A difference between the most and the least common element in the polymer after 10 steps of the process.
- **Data**: Polymer template and the pair insertion rules.
- **Condition**:
  - The insertions all happen simultaneously.
  - Inserted elements are not considered to be part of a pair until the next step.
- **Plan**:
  - Implement the grammar algorithm

In [ ]:
example = """NNCB

CH -> B
HH -> N
CB -> H
NH -> C
HB -> C
HC -> B
HN -> C
NN -> C
BH -> H
NC -> B
NB -> B
BN -> B
BB -> N
BC -> B
CC -> N
CN -> C"""

In [ ]:
def parse(input):
  g = input.split('\n')
  s = g[0]
  r = {}
  for i in range(2, len(g)):
    k,v = g[i].split(' -> ')
    r[k] = v
  return s, r
s, r = parse(example)
print(s)
print(r)

NNCB
{'CH': 'B', 'HH': 'N', 'CB': 'H', 'NH': 'C', 'HB': 'C', 'HC': 'B', 'HN': 'C', 'NN': 'C', 'BH': 'H', 'NC': 'B', 'NB': 'B', 'BN': 'B', 'BB': 'N', 'BC': 'B', 'CC': 'N', 'CN': 'C'}


In [ ]:
r = requests.get(path+'AoC2021_14.txt')
s, rules = parse(r.text[:-1])
print(s)
print(len(rules), rules['KC'])

SHHNCOPHONHFBVNKCFFC
100 B


In [ ]:
def polymerize(p, r):
  """Apply rules `r` to polymer `p`."""
  p0 = list(p)
  # Initialize the new polymer
  p1 = [p0[0]]
  for i in range(len(p0)-1):
    p1.append(r.get(''.join([p0[i],p0[i+1]]),''))
    p1.append(p0[i+1])
  return "".join(p1)

def solve_part1(data):
  p,r = parse(data)

  # Polymerize 10 times
  for i in range(10):
    p = polymerize(p,r)
  
  # Count elements
  c = Counter(p)
  # Sort elements
  e = sorted(c.items(), key=lambda x: x[1])
  # Return the difference
  return e[-1][1]-e[0][1]

solve_part1(example)

1588

In [ ]:
solve_part1(r.text[:-1])

2549

### Part II
- **Unknown**: A difference between the most and the least common element in the polymer after 40 steps of the process.
- **Data**: Same as in the part I.
- **Condition**:
  - Same as i the part I.
- **Plan**:
  - 


In [ ]:
def solve_part2(data):
  p,r = parse(data)

  # Polymerize 40 times
  for i in range(40):
    p = polymerize(p,r)
  
  # Count elements
  c = Counter(p)
  # Sort elements
  e = sorted(c.items(), key=lambda x: x[1])
  # Return the difference
  return e[-1][1]-e[0][1]

solve_part2(example)

In [ ]:
# Relace vyčerpala veškerou RAM a selhala
# 2188189693529

## [Day 13](https://adventofcode.com/2021/day/13): Transparent Origami
### Part I
- **Unknown**: The number of dots visible after completing just the first fold instruction on the transparent paper.
- **Data**: Coordinates of dots and folding instructions
- **Condition**:
  - [0,0] represents the top-left coordinates
  - `x` increases to the right, `y` increases downward
  - `fold along y=...` instruction means folding the paper up
  - `fold along x=xxx` instruction means folding the paper left
- **Plan**:
  - Initialize the matrix
  - Write a function for folding
  - Count the dots after folding


In [ ]:
example = """6,10
0,14
9,10
0,3
10,4
4,11
6,0
6,12
4,1
0,13
10,12
3,4
3,0
8,4
1,10
2,14
8,10
9,0

fold along y=7
fold along x=5"""

In [ ]:
def parse(input):
  lines = input.split('\n')
  dots = []
  folds = []
  for line in lines:
    if ',' in line:
      dots.append(np.array(line.split(',')).astype(int))
    if '=' in line:
      fold = re.findall('[x,y]=\d+', line)[0].split('=')
      folds.append([fold[0],int(fold[1])])
  return np.array(dots), folds

dots, folds = parse(example)
print(dots[:4])
print(folds)

[[ 6 10]
 [ 0 14]
 [ 9 10]
 [ 0  3]]
[['y', 7], ['x', 5]]


In [ ]:
r = requests.get(path+'AoC2021_13.txt')
dots, folds = parse(r.text)
print(dots[-2:])
print(folds[-2:])

[[954 841]
 [960 278]]
[['y', 13], ['y', 6]]


In [ ]:
pap = np.zeros(dots.max(axis=0)+1).astype(int)
for dot in dots:
  pap[dot[0], dot[1]] = 1
pap

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

**Beware** of slicing vs. indexing. The expression `pap[6,10]` adresses one element [6,10] in the array, whereas the expression `pap[[6,10]]` addresses two rows [6, :] and [10, :]

In [ ]:
print(dots[0])
print(pap[dots[0]])
print(np.stack((pap[6, :], pap[10, :])))

[309 320]
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 1]]


In [ ]:
# Fold the paper up - indices
y = 7
for i in range(y):
  print(i, 2*y-i)

0 14
1 13
2 12
3 11
4 10
5 9
6 8


In [ ]:
pap_y7 = pap[:,:y].copy()
for i in range(y):
  pap_y7[:,i] += pap[:, 2*y-i]
pap_y7.T

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
pap_y7[pap_y7>0].shape

(12,)

In [ ]:
# Fold the paper left
x = 5
pap_x5 = pap_y7[:x,:].copy()
for i in range(x):
  pap_x5[i,:] += pap_y7[2*x-i, :]
pap_x5.T

array([[0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0]])

In [ ]:
pap_x5[pap_x5>0].shape

(0,)

In [ ]:
def solve_part1(input):
  dots, folds = parse(input)
  # Initialize the paper
  pap = np.zeros(dots.max(axis=0)+1).astype(int)
  for dot in dots:
    pap[dot[0], dot[1]] = 1
  print('Initial shape:', pap.shape)
  
  # Calculate the first fold
  ax, val = folds[0][0], folds[0][1]
  print(ax, val)
  # Fold the paper left
  if ax == 'x':
    for i in range(val):
      pap[i, :] += pap[2*val-i, :]
    # Adjust shape of the paper
    pap = pap[:val, :]
  else:
    for i in range(val):
      pap[:, i] += pap[:, 2*val-i]
    # Adjust shape of the paper
    pap = pap[:, :val]
  print('Folded shape:', pap.shape)

  return pap[pap>0].shape[0]

solve_part1(example)

Initial shape: (11, 15)
y 7
Folded shape: (11, 7)


17

In [ ]:
solve_part1(r.text)

Initial shape: (1311, 890)
x 655
Folded shape: (655, 890)


669

### Part II
- **Unknown**: Eight capital letters after folding the paper.
- **Data**: Same as in part I
- **Condition**:
  - Same as in part I
- **Plan**:
  - Calculate all folds and return transposed result


In [ ]:
def solve_part2(input):
  dots, folds = parse(input)
  # Initialize the paper
  #pap = np.zeros(dots.max(axis=0)+1).astype(int)
  # Solve IndexError: index 894 is out of bounds for axis 1 with size 890
  pap = np.zeros((1311, 895)).astype(int)
  for dot in dots:
    pap[dot[0], dot[1]] = 1
  print('Initial shape:', pap.shape)
  
  # Calculate the folds
  for ax, val in folds:
    #print(ax, val)
    # Fold the paper left
    if ax == 'x':
      for i in range(val):
        pap[i, :] += pap[2*val-i, :]
      # Adjust shape of the paper
      pap = pap[:val, :]
    else:
      for i in range(val):
        pap[:, i] += pap[:, 2*val-i]
      # Adjust shape of the paper
      pap = pap[:, :val]
    #print('Folded shape:', pap.shape)

  return pap.T

code = solve_part2(example)
print(code)

Initial shape: (1311, 895)
[[1 1 1 1 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 2 2 1 1]
 [0 0 0 0 0]
 [0 0 0 0 0]]


In [ ]:
code = solve_part2(r.text)
print(code)

Initial shape: (1311, 895)
[[10  0  0  7  0  2 13 37  7  0  1  1  1  7  0 26  2 20 31  0  0  1 16  0
   0  2  0  0 13  0  0  6 12  0  0  0  0  2  5  0]
 [12  0  0  5  0 10  0  0  0  0  8  0  0  0  0  0  0  0  1  0  8  0  0  1
   0  5  0  0  9  0 13  0  0  4  0  0  0  0 11  0]
 [ 4  0  0 16  0 24  5  8  0  0 21 17  3  0  0  0  0  9  0  0  3  0  0  0
   0  3  0  0 17  0  2  0  0  0  0  0  0  0  2  0]
 [ 1  0  0 14  0 12  0  0  0  0 19  0  0  0  0  0  7  0  0  0  3  0  0  0
   0  9  0  0  3  0 21  0  0  0  0  0  0  0  1  0]
 [ 5  0  0 15  0  8  0  0  0  0  3  0  0  0  0  2  0  0  0  0 27  0  0  4
   0 18  0  0  4  0  1  0  0 17  0  8  0  0 17  0]
 [ 0 20  2  0  0  1 10 13 15  0 10  0  0  0  0  6  6  3  2  0  0  1  4  0
   0  0  5 19  0  0  0  4  4  0  0  0  8  7  0  0]]


In [ ]:
code =  np.piecewise(code, [code==0, code>0], [0,1]).astype(str)
code = np.piecewise(code, [code == '0', code == '1'], [' ', '#'])
for j in range(5,41,5):
  print(code[:,j-5:j-1])
  print()

[['#' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 [' ' '#' '#' ' ']]

[['#' '#' '#' '#']
 ['#' ' ' ' ' ' ']
 ['#' '#' '#' ' ']
 ['#' ' ' ' ' ' ']
 ['#' ' ' ' ' ' ']
 ['#' '#' '#' '#']]

[['#' '#' '#' '#']
 ['#' ' ' ' ' ' ']
 ['#' '#' '#' ' ']
 ['#' ' ' ' ' ' ']
 ['#' ' ' ' ' ' ']
 ['#' ' ' ' ' ' ']]

[['#' '#' '#' '#']
 [' ' ' ' ' ' '#']
 [' ' ' ' '#' ' ']
 [' ' '#' ' ' ' ']
 ['#' ' ' ' ' ' ']
 ['#' '#' '#' '#']]

[[' ' '#' '#' ' ']
 ['#' ' ' ' ' '#']
 ['#' ' ' ' ' ' ']
 ['#' ' ' ' ' ' ']
 ['#' ' ' ' ' '#']
 [' ' '#' '#' ' ']]

[['#' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 [' ' '#' '#' ' ']]

[[' ' '#' '#' ' ']
 ['#' ' ' ' ' '#']
 ['#' ' ' ' ' ' ']
 ['#' ' ' ' ' ' ']
 ['#' ' ' ' ' '#']
 [' ' '#' '#' ' ']]

[[' ' ' ' '#' '#']
 [' ' ' ' ' ' '#']
 [' ' ' ' ' ' '#']
 [' ' ' ' ' ' '#']
 ['#' ' ' ' ' '#']
 [' ' '#' '#' ' ']]



Result: UEFZCUCJ